BUG: fix union_indexes not supporting sort=False for Index subclasses

AlexKirko opened the pull request:
- [X] closes #35092
- [X] 1 tests added / 1 passed
- [X] passes `black pandas`
- [X] passes `git diff upstream/master -u -- "*.py" | flake8 --diff`
- [X] whatsnew entry

# Scope of PR
The reason for `DataFrame.append` unnecessarily sorting columns turned out to be that when merging indexes we were not passing the `sort` argument for an `Index` subclass. This PR fixes this bug.

# Details
At some point down the call stack, `DataFrame.append` calls `union_indexes` in `core.indexes.api`. In it we detect the type of our indexes: 
```python
indexes, kind = _sanitize_and_check(indexes)
```
We don't pass `sort` for `kind == 'special'` in `union_indexes`.

I had some other ideas, but what I ended up doing is to simply pass the `sort` argument to `Index.union` for `kind == 'special'`. We do have to ignore the `sort` argument for `[MultiIndex, RangeIndex, DatetimeIndex, CategoricalIndex]` to avoid breaking tests and backward compatibility, but it doesn't make sense for these ones to not be sorted when joined anyway, in my opinion.

# Performance
There are no changes to algorithms called. I still did some benchmarking to compare with my previous approach. The current solution didn't have any negative performance impact.

AlexKirko committed:
BUG: fix index detection in _sanitize_and_check

AlexKirko committed:
DOC: add comment to fix

AlexKirko committed:
TST: add tests

AlexKirko committed:
DOC: add whatsnew entry

AlexKirko committed:
REFACT: refact the boolean expression

AlexKirko committed:
REFACT: use generator comprehension

AlexKirko commented:
Hm. There are side-effects, it seems. When the first index is a subset of the second, the extra elements of the second come in their original order now. I'll debug this later.

AlexKirko committed:
TST: add cases to the test

AlexKirko committed:
preserve Index subclass, exclude MultiIndex

AlexKirko commented on pandas/core/indexes/api.py:
```diff
@@ -204,7 +204,16 @@ def conv(i):
+
+        return ind_class(
+            lib.fast_unique_multiple_list([conv(i) for i in inds], sort=sort)
+        )
```
If this is to work, we need to preserve the argument type, otherwise Index subclasses turn into Index, and everything breaks.

AlexKirko replied:
Changed the approach completely.

AlexKirko commented:
Researching. Fiddling with `pd.concat` and `union_indexes` affects a bunch of stuff. Let's see if I can get this bugfix to work without introducing undesired behavior elsewhere.

AlexKirko committed:
include RangeIndex in exceptions

AlexKirko commented:
The current idea is to support `sort` only for `Index` subclasses that are just an `Index` of a specific datatype.

AlexKirko committed:
exclude DatetimeIndex

DatetimeIndex has an explicit branch set up for it with kind == 'special'

AlexKirko committed:
exclude CategoricalIndex

AlexKirko committed:
CLN: sort imports in test_common

AlexKirko committed:
DOC: remove unnecessary backquoutes in whatsnew

AlexKirko committed:
DOC: add missing colon to whatsnew

AlexKirko committed:
pass sort to Index.union instead of previous approach

AlexKirko committed:
explicitly ignore sort for heavily modified Index subclasses

If an Index isn't just a sliceable one-level array (Index), we should
not mess with sort to avoid breaking backwards compatibility.

AlexKirko commented on pandas/core/indexes/api.py:
```diff
@@ -206,6 +206,17 @@ def conv(i):
+    ):
+        ignore_sort = True
+    else:
+        ignore_sort = False
```
If we let any of these index type not get sorted, we'll break some tests where we assume that the results of joining on any of these come out sorted. Moreover, there isn't much point to having any of them unsorted except for pretty printing purposes, in my opinion.
If you think we should pass sort through for every type of `Index` subclass, please let me know.

jreback replied:
can you show what tests break?

jreback replied:
i would rather just pass thru and adjust the tests

AlexKirko replied:
done

AlexKirko committed:
fully incorporate sort ignoring for particular Index subtypes

AlexKirko committed:
CLN: fix typo

AlexKirko commented:
@gfyoung This PR is ready for a review now.

gfyoung commented on pandas/core/indexes/api.py:
```diff
@@ -206,6 +206,17 @@ def conv(i):
 
+    # GH 35092. Detect if we have an Index type for which the sort
+    # setting doesn't make sense
+    ind_types = list({type(index) for index in indexes})
```
Is this cast to `list` necessary?

AlexKirko replied:
@gfyoung 
Sorry, copied it from lower in the code without thinking. An ordinary list comprehension will suffice.

gfyoung commented on pandas/core/indexes/api.py:
```diff
@@ -206,6 +206,17 @@ def conv(i):
+    # setting doesn't make sense
+    ind_types = list({type(index) for index in indexes})
+    if any(
+        ind_type in [MultiIndex, RangeIndex, DatetimeIndex, CategoricalIndex]
```
Please add tests for all of these types that would trigger `ignore_sort=True`

AlexKirko replied:
@gfyoung 
Re-tested two at a time (old tests were done with a different bugfix implementation, and it confounded the results):

* `MultiIndex` breaks `test_unbalanced` in `ests.reshape.test_melt.TestWideToLong` . From what I can see, we end up not resorting by `id` (first index level) and are instead left with the row order of `year` (the initial `DataFrame`). `MultiIndex` also breaks `test_nonnumeric_suffix` and `test_float_suffix` in `tests.reshape.test_melt.TestWideToLong`. Conceptually, `melt` has the same problem `wide_to_long`.
* `RangeIndex` breaks `test_str_cat_align_mixed_inputs` in `tests.test_strings.TestStringMethods`. There is a indexed/unindexed case and having a default RangeIndex as one of the indices sorts the result. 
* `DatetimeIndex` is safe.
* `CategoricalIndex` breaks `test_construct_with_two_categoricalindex_series` in `tests.frame.test_constructors.TestDataFrameConstructors`. The test assumes that the dataframe resulting from joining two categorical Series will have the categories sorted.

I'm up for changing the tests if you and @jreback think that would be better.

AlexKirko committed:
CLN: switch ind_types to list comprehension

AlexKirko committed:
test without DatetimeIndex and CategoricalIndex in exceptions

AlexKirko committed:
test without multi, range, datetime in exceptions

AlexKirko committed:
return MultiIndex, RangeIndex, CategoricalIndex to exceptions

AlexKirko committed:
CLN: black indexes/api.py

AlexKirko committed:
Merge branch 'master' into append-err-sort

AlexKirko commented:
Web and docs seem to be failing for everyone atm.

AlexKirko committed:
restart tests

AlexKirko committed:
Merge branch 'master' into append-err-sort

AlexKirko commented:
CI / Checks now fail because isort dropped the `--recursive`` key. Submitted an issue #35134

AlexKirko committed:
Merge branch 'master' into append-err-sort: pin isort and sphinx

AlexKirko committed:
add DatetimeIndex back into exceptions

DatetimeIndex seems to break test_basic_series_frame_alignment
in test_eval.py

AlexKirko commented on pandas/core/indexes/api.py:
```diff
@@ -206,6 +206,17 @@ def conv(i):
+    # setting doesn't make sense
+    ind_types = [type(index) for index in indexes]
+    if any(
+        ind_type in [MultiIndex, RangeIndex, DatetimeIndex, CategoricalIndex]
```
@gfyoung Turns out that `DatetimeIndex` breaks using eval to add a `Series` and a `DataFrame` in  `test_basic_series_frame_alignment` in `test_eval.py` but only on the [MacOS pipeline](https://dev.azure.com/pandas-dev/pandas/_build/results?buildId=38693&view=logs&j=bef1c175-2c1b-51ae-044a-2437c76fc339&t=770e7bb1-09f5-5ebf-b63b-578d2906aac9&l=105). So I'm keeping this in the exception list too.

This is the last example of each of these types of indices breaking sorting expectations somewhere.

jreback commented on doc/source/whatsnew/v1.1.0.rst:
```diff
@@ -976,6 +976,7 @@ Indexing
 - Bug in :meth:`DataFrame.loc` with dictionary of values changes columns with dtype of ``int`` to ``float`` (:issue:`34573`)
 - Bug in :meth:`Series.loc` when used with a :class:`MultiIndex` would raise an IndexingError when accessing a None value (:issue:`34318`)
 - Bug in :meth:`DataFrame.reset_index` and :meth:`Series.reset_index` would not preserve data types on an empty :class:`DataFrame` or :class:`Series` with a :class:`MultiIndex` (:issue:`19602`)
+- Bug in :func:`pandas.core.indexes.api.union_indexes` would lead to :meth:`DataFrame.append` sorting columns even when ``sort=False`` is specified (:issue:`35092`)
```
you don't need the first part (as this is a private function); move to reshaping section of Bug Fixes

jreback commented on pandas/tests/indexes/test_common.py:
```diff
@@ -395,3 +396,15 @@ def test_astype_preserves_name(self, index, dtype, copy):
             assert result.name == index.name
+
+
+@pytest.mark.parametrize("exp_arr, sort", [([0, 1, 4, 3], False), ([0, 1, 3, 4], True)])
```
can you use the index fixture

AlexKirko replied:
Done. Using the `sort` fixture. There are only two in conftest, and only this one is useful.

jreback reviewed (changes requested):
can you add the test from the OP (.append), and all cases of it.

AlexKirko committed:
DOC: move to Reshaping in whatsnew and edit

AlexKirko committed:
pass sort through to all Index subclasses

AlexKirko committed:
TST: alter test_construct_with_two_categoricalindex_series

AlexKirko committed:
TST: alter test_str_cat_align_mixed_inputs

AlexKirko committed:
TST: alter test_unbalanced in test_melt

AlexKirko committed:
TST: alter nonnumeric and float suffix tests in test_melt

AlexKirko committed:
CLN: run black pandas

AlexKirko committed:
TST: add OP test, use fixture in the index test

pep8speaks commented:
Hello @AlexKirko! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:







* In the file [`pandas/tests/reshape/test_concat.py`](https://github.com/pandas-dev/pandas/blob/0479618ad82cbbe71e806a383f1757c5dbb2c6c3/pandas/tests/reshape/test_concat.py):

> [Line 2865:35](https://github.com/pandas-dev/pandas/blob/0479618ad82cbbe71e806a383f1757c5dbb2c6c3/pandas/tests/reshape/test_concat.py#L2865): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 2865:45](https://github.com/pandas-dev/pandas/blob/0479618ad82cbbe71e806a383f1757c5dbb2c6c3/pandas/tests/reshape/test_concat.py#L2865): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 2866:35](https://github.com/pandas-dev/pandas/blob/0479618ad82cbbe71e806a383f1757c5dbb2c6c3/pandas/tests/reshape/test_concat.py#L2866): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','
> [Line 2866:45](https://github.com/pandas-dev/pandas/blob/0479618ad82cbbe71e806a383f1757c5dbb2c6c3/pandas/tests/reshape/test_concat.py#L2866): [E231](https://duckduckgo.com/?q=pep8%20E231) missing whitespace after ','

AlexKirko committed:
CLN: run black on test files

AlexKirko commented:
@jreback Changes made, ready for another review:
1. Pass sort through to all `Index` subclasses. Alter tests to accomodate no automatic sorting. Add comments to changes where appropriate.
2. Add test for `append` to `reshape/test_concat.py`. Use streamlined equivalent of the OP's test case, as the original case is overly verbose.
3. Move whatsnew record to reshape section and remove mention of the private function.

jreback commented on pandas/core/indexes/api.py:
```diff
@@ -214,7 +214,11 @@ def conv(i):
+                # Index.union expects sort=None instead of sort=True
+                if sort:
+                    sort = None
+                result = result.union(other, sort=sort)
```
really? why is this

jreback replied:
IIRC sort is implemented for all union index types, or is it showing a warning?

AlexKirko replied:
Ah, sorry, this confused me:

```
        sort : bool or None, default None
            Whether to sort the resulting Index.

            * None : Sort the result, except when

              1. `self` and `other` are equal.
              2. `self` or `other` has length 0.
              3. Some values in `self` or `other` cannot be compared.
                 A RuntimeWarning is issued in this case.

            * False : do not sort the result.
```
Works fine with `True`. Pushing the fix.

jreback replied:
hmm IIRC we did this deliberartly as we don't support this, can you see if can add an existing issue reference here (in both places if possible). future us will be puzzzled otherwise.

AlexKirko replied:
@jreback 
Did a little digging. The API was changed in issue #24959. `sort` had been introduced to all the Index set-operations before, and then we discovered that the operations didn't guarantee sorting in edge cases. We decided to change the default to `sort=None` for set-ops to signify that and fix this in the future: ensure that `sort=True` always sorts for every set-op and then change the default to `sort=False` and deprecate `sort=None`. As far as I can tell, the last attempt to implement `sort=True` and deprecate `sort=None` was in #25980, and the PR went stale, so we never finished the job.

While we can and do hit edge cases in `Index.union`, there is a crutch in `_get_combined_index` that saves us if `union` doesn't sort:

```python
    if sort:
        try:
            index = index.sort_values()
        except TypeError:
            pass
```

I've added a comment to this PR to clarify all this stuff.

AlexKirko committed:
REFACT: remove unnecessary change of sort from True to None

AlexKirko committed:
Revert "REFACT: remove unnecessary change of sort from True to None"

This reverts commit ada734694d477ae5000367f2250929f0b9297afb. It threw an error.

AlexKirko committed:
restart tests

AlexKirko commented:
All green.

AlexKirko committed:
DOC: add comment with ref to sort=None issue

sort=None signifies that Index.union doesn't gurarantee sorting
or error by default. It might mirror legacy behavior and leave
the Index unsorted in edge cases.

AlexKirko committed:
DOC: clarify comment language

AlexKirko committed:
DOC: clarify comment more

jreback reviewed (approved):
lgtm ping on green.

AlexKirko committed:
restart tests

AlexKirko commented:
@jreback 
All green. There was some hardware incident that impacted Azure pipelines, but we are good now.

jreback commented:
thanks very nice!
